src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -545,137 +571,319 @@ private RegexNode ReduceSet()
+            //   e.g. "apple|(?:orange|pear)|grape" -> "apple|orange|pear|grape"
+            void ReduceSingleLetterAndNestedAlternations()
             {
-                at = children[i];
+                bool wasLastSet = false;
+                bool lastNodeCannotMerge = false;
+                RegexOptions optionsLast = 0;
+                RegexOptions optionsAt;
+                int i;
+                int j;
+                RegexNode at;
+                RegexNode prev;
+
+                List<RegexNode> children = (List<RegexNode>)Children!;
+                for (i = 0, j = 0; i < children.Count; i++, j++)
+                {
+                    at = children[i];
 
-                if (j < i)
-                    children[j] = at;
+                    if (j < i)
+                        children[j] = at;
 
-                while (true)
-                {
-                    if (at.Type == Alternate)
+                    while (true)
                     {
-                        if (at.Children is List<RegexNode> atChildren)
+                        if (at.Type == Alternate)
                         {
-                            for (int k = 0; k < atChildren.Count; k++)
+                            if (at.Children is List<RegexNode> atChildren)
+                            {
+                                for (int k = 0; k < atChildren.Count; k++)
+                                {
+                                    atChildren[k].Next = this;
+                                }
+                                children.InsertRange(i + 1, atChildren);
+                            }
+                            else
                             {
-                                atChildren[k].Next = this;
+                                RegexNode atChild = (RegexNode)at.Children!;
+                                atChild.Next = this;
+                                children.Insert(i + 1, atChild);
                             }
-                            children.InsertRange(i + 1, atChildren);
+                            j--;
                         }
-                        else
+                        else if (at.Type == Set || at.Type == One)
                         {
-                            RegexNode atChild = (RegexNode)at.Children!;
-                            atChild.Next = this;
-                            children.Insert(i + 1, atChild);
-                        }
-                        j--;
-                    }
-                    else if (at.Type == Set || at.Type == One)
-                    {
-                        // Cannot merge sets if L or I options differ, or if either are negated.
-                        optionsAt = at.Options & (RegexOptions.RightToLeft | RegexOptions.IgnoreCase);
+                            // Cannot merge sets if L or I options differ, or if either are negated.
+                            optionsAt = at.Options & (RegexOptions.RightToLeft | RegexOptions.IgnoreCase);
```

Comment:
RightToLeft cannot be set inline. So all of the pattern should have the same value for it, and a test is unnecessary?

https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-options

These can be all set inline
```
RegexOptions.IgnoreCase
RegexOptions.Multiline
RegexOptions.Singleline
RegexOptions.ExplicitCapture
RegexOptions.IgnorePatternWhitespace
```
(Might be pretty hard to grok such a pattern..)

Do we need to check all these? At every point we check nodes for equivalence? (Again ideally there would be some way to do this centrally in a helper)

Reply:
> RightToLeft cannot be set inline

It can't be the developer, but it does get set by the implementation for lookbehinds.  I doubt that would affect this code, but I'm hesitant to go and change it, especially as part of this PR.  This PR isn't adding this code, it's only changing the indentation as part of moving it into a helper function.